pkg/services/sqlstore/session/session.go

```diff
@@ -16,26 +19,74 @@ type Session interface {
 
 type SessionDB struct {
 	sqlxdb *sqlx.DB
+	logger sqllog.ILogger
 }
 
-func GetSession(sqlxdb *sqlx.DB) *SessionDB {
-	return &SessionDB{sqlxdb: sqlxdb}
+func GetSession(sqlxdb *sqlx.DB, debugSQL bool) *SessionDB {
+	db := SessionDB{sqlxdb: sqlxdb}
+	if !debugSQL {
+		db.logger = sqllog.DiscardLogger{}
+	} else {
+		db.logger = sqllog.NewGenericLogger(glog.LvlInfo, glog.WithSuffix(glog.New("sqlstore.sqlx"), glog.CallerContextKey, glog.StackCaller(glog.DefaultCallerDepth)))
+
+	}
+	return &db
+}
+
+type DBFunc func(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+type DBFuncWithResult func(ctx context.Context, query string, args ...interface{}) (sql.Result, error)
+type DBFuncWithOneArgAndResult func(ctx context.Context, query string, arg interface{}) (sql.Result, error)
+
+func logDBFunc(logger sqllog.ILogger, callback DBFunc) DBFunc {
+	return func(ctx context.Context, dest interface{}, query string, args ...interface{}) error {
+		start := time.Now()
+		err := callback(ctx, dest, query, args...)
+		if err != nil {
+			logger.Error(err)
+		}
+		logger.Infof("[SQL] %v %v - took: %v", query, args, time.Since(start))
+		return err
+	}
+}
+
+func logDBFuncWithResult(logger sqllog.ILogger, callback DBFuncWithResult) DBFuncWithResult {
+	return func(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
+		start := time.Now()
+		r, err := callback(ctx, query, args...)
+		if err != nil {
+			logger.Error(err)
+		}
+		logger.Infof("[SQL] %v %v - took: %v", query, args, time.Since(start))
+		return r, err
+	}
+}
+
+func logDBFuncWithArgAndResult(logger sqllog.ILogger, callback DBFuncWithOneArgAndResult) DBFuncWithOneArgAndResult {
+	return func(ctx context.Context, query string, arg interface{}) (sql.Result, error) {
+		start := time.Now()
+		r, err := callback(ctx, query, arg)
+		if err != nil {
+			logger.Error(err)
+		}
+		logger.Infof("[SQL] %v %v - took: %v", query, arg, time.Since(start))
+		return r, err
+	}
```

Comment:
Ideally I would like to use the same wrapper function but it's hard (if not impossible) because the inner functions have different signatures.
Any better suggestions are welcome!
Maybe an option would be using reflection but I'm not sure we would like to do that.